### ISSUE TITLE
Memory leak with callback registered to IoTHubClient 

### ISSUE THREAD

**Issue by alex007sirois** [Azure/azure-iot-sdk-python#181]
- **OS and version used:** <VERSION>
Raspberry Pi 3 Model B+ 
Raspbian GNU/Linux 9 (stretch)
Linux raspberrypi 4.14.34-v7+ #1110 SMP Mon Apr 16 15:18:51 BST 2018 armv7l GNU/Linux

- **Python runtime used:** <VERSION>
Python 2.7.13

- **SDK version used:** <VERSION>
azure-iothub-device-client==1.4.2
libboost1.62-all

# Description of the issue:
Instanciating a IoTHubClient and then registering any callback (receive_message, device_twin, connection_status, device_method) will result in a memory leak. Those callback stay in memory even if you "close" the IoTHubClient (there is no close method, but removing references to the client will clear it from memory). So if the function you passed is in fact a method of a class instance, that class instance will stay in memory until the python process ends. The reference seems to be coming from the C as the garbage collector cannot find the referrers.

To prove my point, I create a IoTHubClient and register a callback on it. I then create a weakref to the callback and the client to then delete the references I have to them. The client is then cleaned from memory, but the callback isn't. The garbage collector cannot find the referrer, but there is 1 ref left to the callback.

# Code sample exhibiting the issue:
```
from iothub_client import IoTHubClient, IoTHubTransportProvider

connection_string = "connection string"
protocol = IoTHubTransportProvider.MQTT
client = IoTHubClient(connection_string, protocol)

def receive_message_callback(message, user_context): pass

client.set_message_callback(receive_message_callback, 0)

import weakref
client_ref = weakref.ref(client)
function_ref = weakref.ref(receive_message_callback)

del client
del receive_message_callback

print(client_ref)
>>> <weakref at 0x76865630; dead>

print(function_ref)
>>> <weakref at 0x7687c720; to 'function' at 0x76871930 (receive_message_callback)>
# The function is still refered somewhere in code, else it would look like the client_ref

import gc
print(gc.get_referrers(function_ref()))
>>> []
# The garbage collector does not know any referrer to the function

import sys
print(sys.getrefcount(function_ref()))
>>> 2
# There are 2 ref to the function where 1 is sys.getrefcount
# In other words, there is 1 ref to our functions that the gc cannot find
# It is most likely in the C sdk through boost
```

I also saw some crashing errors that might be linked to this callback cleaning problem

```
Fatal Python error: GC object already tracked
Aborted
```

This error stops the whole process and cannot be catched in python, so it is most likely a problem in a C lib and searching around the internet points to this specific error being because of boost not cleaning callbacks (seems familiar)

**Comment by pierreca** [comment]
@alex007sirois thanks for reporting this - pinging @jspaith from the C SDK to see what he thinks about this.

**Comment by alex007sirois** [comment]
Biggest problem to me is if the callback is an instance's method, the instance will stay in memory. We don't really have a problem with this, but this could hurt in a project where this instance really needs to get out of memory.

Since we know of this problem, we can work around it so it does not cause much problem on our part. But maybe there should be a warning somewhere in doc or something?

**Comment by Ram9293** [comment]
@alex007sirois Hi Alex, I am facing the same issue of python crashing in Production evironment, we are sending commands on an interval of every 15mins, after sometime python crashes. May I know the workaround you followed?

Thanks in advance.

@pierreca If you can also let us know a best possible workaround, it would be great. Thanks in advance

**Comment by jebrando** [comment]
With the Python v2 reaching General Availability we are recommending users move away from the v1 SDK.  We will be closing the v1 SDK issues whose features are covered by the v2 release.  Feel free to open an new issue if you find an issue with the v2 SDK.

Thank you for your patronage

**Comment by az-iot-builder-01** [comment]
@alex007sirois, @Ram9293, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
